### PR TITLE
Iox-#1345 Cleanup logger folder structure and add cmake option for minimal log level

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR adds a cmake parameter to set the minimal log level which will be compiled into the binary. All log messages with a lower log level will be optimized away during compilation. This adds the option to completely get rid of the log calls if the log level is set to `OFF`.

In addition, some cleanups where done regarding the folder structure since the previous idea with moving some parts to iceoryx_platform didn't pan out well.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1345

### CODE REVIEW

File: iceoryx_hoofs/BUILD.bazel [hunk 1]
```diff
@@ -15,14 +15,27 @@
...
+
+configure_file(
+    name = "iceoryx_hoofs_deployment_hpp",
+    src = "cmake/iceoryx_hoofs_deployment.hpp.in",
+    out = "include/iceoryx_hoofs/iceoryx_hoofs_deployment.hpp",
+    config = {
+        # FIXME: for values see "iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake" ... for now some nice defaults
+        "IOX_MINIMAL_LOG_LEVEL": "TRACE",
+    },
+)
```
Comment (elBoberido) [inline iceoryx_hoofs/BUILD.bazel:28]: @dkroenke I hope this is the idiomatic way

File: iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+message(STATUS "[i] <<<<<<<<<<<<< Start iceoryx_posh configuration: >>>>>>>>>>>>>")
```
Comment (mossmaurice) [inline iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake:18]:
```suggestion
message(STATUS "[i] <<<<<<<<<<<<< Start iceoryx_hoofs configuration: >>>>>>>>>>>>>")
```

File: iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+if(NOT IOX_MINIMAL_LOG_LEVEL)
+    set(IOX_MINIMAL_LOG_LEVEL "FATAL")
+endif()
+message(STATUS "[i] IOX_MINIMAL_LOG_LEVEL: " ${IOX_MINIMAL_LOG_LEVEL})
+
+message(STATUS "[i] <<<<<<<<<<<<<< End iceoryx_posh configuration: >>>>>>>>>>>>>>")
```
Comment (mossmaurice) [inline iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake:25]:
```suggestion
message(STATUS "[i] <<<<<<<<<<<<<< End iceoryx_hoofs configuration: >>>>>>>>>>>>>>")
```

File: iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+# Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
```
Comment (mossmaurice) [inline iceoryx_hoofs/cmake/IceoryxHoofsDeployment.cmake:2]:
```suggestion
# Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_ICEORYX_HOOFS_TYPES_INL
+#define IOX_HOOFS_ICEORYX_HOOFS_TYPES_INL
+
+#include "iceoryx_hoofs/iceoryx_hoofs_types.hpp"
+
+/// @note since this file will be included by many other files, it should not include other header except
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl:22]: I like this comment!

**Review by mossmaurice — COMMENTED** [review]
LGTM, only some minor findings.

File: iceoryx_hoofs/CMakeLists.txt [hunk 6]
```diff
@@ -82,6 +85,15 @@ iox_add_library(
         source/units/duration.cpp
 )
 
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/cmake/iceoryx_hoofs_deployment.hpp.in"
+  "${CMAKE_BINARY_DIR}/generated/iceoryx_hoofs/include/iceoryx_hoofs/iceoryx_hoofs_deployment.hpp" @ONLY)
+
+install(
```
Comment (elfenpiff) [inline iceoryx_hoofs/CMakeLists.txt:91]: This is not required anymore. `iox_add_library` and `iox_add_executable` have an install directive which take care of the right prefix, project name etc.
I admit that they do not yet install generated hpp file (so it seems at least when I look at the code) but they seem to install all interface files which this file should be part of it.

So could you remove this install directive and check if the file is not already installed by `iox_add_library`. If not could you add an additional argument `INSTALL_ADDITIONAL_FILES` where this file is provided. Then you could add the correct prefix inside the `iox_add_library` thingy.
Whenever we decide to restructure this we only have to touch `iox_add_library` and nothing else but when there is an `install` directive outside it complicates things again.

File: iceoryx_hoofs/include/iceoryx_hoofs/iceoryx_hoofs_types.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+namespace iox
+{
+namespace cxx
+{
+using byte_t = uint8_t;
+}
+
+namespace log
+{
+enum class LogLevel : uint8_t
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/iceoryx_hoofs_types.hpp:34]: Oh no a public type without a doxygen description. :smile:

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        return "LogLevel::WARN";
+    case LogLevel::INFO:
+        return "LogLevel::INFO";
+    case LogLevel::DEBUG:
+        return "LogLevel::DEBUG";
+    case LogLevel::TRACE:
+        return "LogLevel::TRACE";
+    }
+
+    return "[Undefined LogLevel]";
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl:49]: Please rephrase it to `[LogLevel::EAT_MY_SHORTS]`

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/log/building_blocks/console_logger.inl [hunk 9]
```diff
@@ -15,14 +15,14 @@
...
-#define IOX_HOOFS_PLATFORM_BUILDING_BLOCKS_CONSOLE_LOGGER_INL
+#ifndef IOX_HOOFS_LOG_BUILDING_BLOCKS_CONSOLE_LOGGER_INL
+#define IOX_HOOFS_LOG_BUILDING_BLOCKS_CONSOLE_LOGGER_INL
 
-#include "iceoryx_hoofs/log/platform_building_blocks/console_logger.hpp"
+#include "iceoryx_hoofs/log/building_blocks/console_logger.hpp"
 
 namespace iox
 {
-namespace pbb
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/log/building_blocks/console_logger.inl:25]: What did `pbb` stand for?

Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/log/building_blocks/console_logger.inl:25]: platform building blocks ... usually I don't want to type that much :)

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl (hunk 8 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/iceoryx_hoofs_types.inl:49]: denied :)

File: iceoryx_hoofs/CMakeLists.txt (hunk 6 above)
Comment (elBoberido) [reply iceoryx_hoofs/CMakeLists.txt:91]: @elfenpiff I copied this from posh. I guess it's currently not possible. I'll have a look at it

Comment (elBoberido) [reply iceoryx_hoofs/CMakeLists.txt:91]: @elfenpiff unfortunately it is not easily doable in a clean way. I have an idea how to do it but I'm not sure if it works and it takes some time. Since we do it the same way in posh, would you be fine if I create an issue for this to solve it in both packages?

Comment (elfenpiff) [reply iceoryx_hoofs/CMakeLists.txt:91]: I am fine with creating an issue. Lets merge this

**Review by elfenpiff — APPROVED** [review]

**Review by mossmaurice — DISMISSED** [review]

File: iceoryx_hoofs/CMakeLists.txt (hunk 6 above)
Comment (elBoberido) [reply iceoryx_hoofs/CMakeLists.txt:91]: Done with #1687

**Review by elfenpiff — DISMISSED** [review]

**Review by elfenpiff — DISMISSED** [review]

**Review by mossmaurice — APPROVED** [review]

**Review by elfenpiff — DISMISSED** [review]
Since @dkroenke is already running into integration issues and trouble since the new logger was introduced but the changes and migration path is missing in our changelog I insist now that we do this in this PR.
For me it is the next logger related PR which gets merged and therefore which have to do it now.
 
A compromise would be a new PR which add the changes to the changelog and is merged before this PR.

**Review by dkroenke — APPROVED** [review]
